### ISSUE TITLE
[BUG] Umlauts (ä ö ü ß) are removed from imported PDF

### ISSUE THREAD

**Issue by ramteid** [paperless-ngx/paperless-ngx#4139]
### Description

# The issue

Umlaut characters (ä ü ö ß) are present in the original PDF but removed in the ORC'd PDF.

# Expected behavior

The processed PDF matches the original.

# Actual behavior

The processed PDF does not contain any umlauts (ä ü ö ß) anymore, whitespaces instead.

# More details

I downloaded my monthly statement of my bank account as a PDF. It's a German bank, so the document contains several umlaut characters (ä ü ö ß). The original PDF contains text which is selectable and copyable. PNGX consumed the PDF from the default consumption folder.

The original PDF:
![image](https://github.com/paperless-ngx/paperless-ngx/assets/710061/9282fb07-ccbe-4f3e-af46-798d5e14e519)

The original still contains all the umlauts.

The PDF how it looks like when I open it via the PNGX web ui or alternatively download it from the web UI:
![image](https://github.com/paperless-ngx/paperless-ngx/assets/710061/1c44ac05-afca-4074-b45d-1541031d4cdc)

The PDF obtained via the PNGX web ui is missing all the umlauts, with whitespaces inserted instead.
I tested without any pre- or postprocessor enabled.

### Steps to reproduce

- Let PNGX consume this document (e. g. via configured folder)
  [umlaut-test.pdf](https://github.com/paperless-ngx/paperless-ngx/files/12560600/umlaut-test.pdf)
- Download the document via the PNGX web UI and download it
- The downloaded PDF should be missing all umlaut / non-ASCII characters

### Webserver logs

```bash
[2023-09-08 16:01:10,019] [INFO] [paperless.management.consumer] Adding /usr/src/paperless/consume/TESTFILE.pdf to the task queue.

[2023-09-08 16:01:10,156] [INFO] [paperless.consumer] Consuming TESTFILE.pdf

[2023-09-08 16:01:10,160] [DEBUG] [paperless.consumer] Detected mime type: application/pdf

[2023-09-08 16:01:10,162] [DEBUG] [paperless.consumer] Parser: RasterisedDocumentParser

[2023-09-08 16:01:10,167] [DEBUG] [paperless.consumer] Parsing TESTFILE.pdf...

[2023-09-08 16:01:10,370] [DEBUG] [paperless.parsing.tesseract] Calling OCRmyPDF with args: {'input_file': PosixPath('/tmp/paperless/paperless-ngx2f13qfpb/TESTFILE.pdf'), 'output_file': PosixPath('/tmp/paperless/paperless-q1_nna88/archive.pdf'), 'use_threads': True, 'jobs': 4, 'language': 'deu', 'output_type': 'pdfa', 'progress_bar': False, 'skip_text': True, 'clean': True, 'deskew': True, 'rotate_pages': True, 'rotate_pages_threshold': 12.0, 'sidecar': PosixPath('/tmp/paperless/paperless-q1_nna88/sidecar.txt')}

[2023-09-08 16:01:11,228] [DEBUG] [paperless.parsing.tesseract] Incomplete sidecar file: discarding.

[2023-09-08 16:01:11,260] [DEBUG] [paperless.consumer] Generating thumbnail for TESTFILE.pdf...

[2023-09-08 16:01:11,264] [DEBUG] [paperless.parsing] Execute: convert -density 300 -scale 500x5000> -alpha remove -strip -auto-orient /tmp/paperless/paperless-q1_nna88/archive.pdf[0] /tmp/paperless/paperless-q1_nna88/convert.webp

[2023-09-08 16:01:13,505] [DEBUG] [paperless.consumer] Saving record to database

[2023-09-08 16:01:13,506] [DEBUG] [paperless.consumer] Creation date from st_mtime: 2023-09-08 16:00:47.056523+02:00

[2023-09-08 16:01:13,889] [INFO] [paperless.handlers] Assigning correspondent CUSTOMBANK to 2023-09-08 TESTFILE

[2023-09-08 16:01:13,898] [INFO] [paperless.handlers] Assigning document type Bescheid to 2023-09-08 CUSTOMBANK TESTFILE

[2023-09-08 16:01:13,916] [INFO] [paperless.handlers] Assigning storage path USER's Dokumente to 2023-09-08 CUSTOMBANK TESTFILE

[2023-09-08 16:01:13,993] [DEBUG] [paperless.filehandling] Document has storage_path 1 (USER/{created_year}/{correspondent}/{title}) set

[2023-09-08 16:01:14,005] [DEBUG] [paperless.filehandling] Document has storage_path 1 (USER/{created_year}/{correspondent}/{title}) set

[2023-09-08 16:01:14,010] [DEBUG] [paperless.consumer] Deleting file /tmp/paperless/paperless-ngx2f13qfpb/TESTFILE.pdf

[2023-09-08 16:01:14,015] [DEBUG] [paperless.parsing.tesseract] Deleting directory /tmp/paperless/paperless-q1_nna88

[2023-09-08 16:01:14,016] [INFO] [paperless.consumer] Document 2023-09-08 CUSTOMBANK TESTFILE consumption finished
```

### Paperless-ngx version

1.17.0

### Host OS

Linux nas 4.4.180+ #42962 SMP Fri May 26 23:30:43 CST 2023 x86_64 GNU/Linux

### Installation method

Docker - official image

### Browser

Chrome

### Configuration changes

No relevant changes

**Comment by stumpylog** [comment]
Just running the command listed in the logs for `ocrmypdf` produces this.  It's probably the case that Tesseract isn't able to detect the accents, which there's nothing we're able to do to fix.  You can try various setting, like perhaps a full redo of OCR would work, instead of skipping text, but overall, we're restricted by what other tools produce.

**Comment by ramteid** [comment]
@stumpylog Thanks for your quick reply. Obviously the error happens in `ocrmypdf`. However, `ocrmypdf` behaves this way only within the paperless-ngx Docker image.

The latest PNGX version 1.17 contains `ocrmypdf` version `14.3.0`. The error can be reproduced by:

`docker run --rm -i --entrypoint ocrmypdf paperlessngx/paperless-ngx --use-threads --jobs 4 --language deu --output-type pdfa --skip-text --clean --desk
ew --rotate-pages --rotate-pages-threshold 12.0 - - <umlaut-test.pdf >output.pdf`
´´´

Using the same `ocrmypdf` version from their official docker image, it works fine (umlauts remain):

```
docker run -i --rm jbarlow83/ocrmypdf:v14.3.0 --use-threads --jobs 4 --language deu --output-type pdfa --skip-text --clean --deskew --rotate-pages --rotate-pages-threshold 12.0 - - <umlaut-test.pdf >output.pdf
´´´

So I suppose there's something wrong with the integration of `ocrmypdf` into paperless-ngx. Maybe you can help me track this down further?

**Comment by stumpylog** [comment]
They have different versions of ghostscript, tesseract and unpaper, all of which likely have some effect, gs and tess the most.  Those versions will change when the OS updates them, beyond our control.

**Comment by Totil-a** [comment]
Why is this beyond your control?
I thought docker allows for controlled environment with all required and correct dependencies?
Just trying to understand.

**Comment by stumpylog** [comment]
OS package repos provide the versions.  They use a different distro, and so different versions in their repos.

**Comment by ramteid** [comment]
@stumpylog Please don't underestimate the ramifications of this issue for European users (and other non-ASCII localized users). With this bug, I have to expect some of my data to be partially **_destroyed_** when importing it into PNGX (!!!) As a result, non-English users cannot use paperless-ngx without concern about their data. 

While it is true you cannot control the OS package repositories, you still could manually include newer versions of ocrmypdf, tesseract, ghostscript, unpaper and such. That won't be trivial of course.

Meanwhile, I helped myself with a kind of ugly workaround: Doing the OCR step using a working version of ocrmypdf (`jbarlow83/ocrmypdf:v14.4.0`) in a pre-consumption script. Inputting the OCR'd PDF resulting from the pre-consumption script into the regular PNGX OCR process ultimately results with all umlauts intact. This might not be future-proof but it seems to be a hot-fix for now. Here's how:

- `cd` to where your `docker-compose.yml` is located

- create a subfolder with the following script file `pre-consumption/ocr-pdf`:
```
#!/bin/bash

# Exit on first error
set -e

echo "Performing Pre-OCR on $DOCUMENT_WORKING_PATH"

# Copy file to a folder shared with the latest ocrmypdf container
FILE_NAME=$(basename "$DOCUMENT_WORKING_PATH")
cp "$DOCUMENT_WORKING_PATH" "/ocrmypdf/$FILE_NAME"

# Rename it after moving to avoid race conditions
mv "/ocrmypdf/$FILE_NAME" "/ocrmypdf/pre-ocr-$FILE_NAME"

# Wait for it to be processed
POST_FILE_NAME="/ocrmypdf/post-ocr-$FILE_NAME"
counter=0
while [ ! -e "$POST_FILE_NAME" ]; do
    if [ "$counter" -ge 30 ]; then
        echo "Timeout reached. Assuming pre-processing OCR failed"
        exit 1
    fi
    echo "Waiting for $POST_FILE_NAME to be present..."
    sleep 3
    counter=$((counter + 3))
done

# Replace the original working copy with the modified PDF
mv "$POST_FILE_NAME" "$DOCUMENT_WORKING_PATH"
```
(possibly chmod 777 it if required)

- In your `docker-compose.yml` modify the service `webserver`:
```
    volumes:
      ...
      - ./pre-consumption/ocr-pdf:/usr/local/bin/ocr-pdf
      - ocrmypdf-share:/ocrmypdf:z
      ...
    environment:
      PAPERLESS_PRE_CONSUME_SCRIPT: "/usr/local/bin/ocr-pdf"
```

- In your `docker-compose.yml` add the service `ocrmypdf`:
```
  ocrmypdf:
    image: jbarlow83/ocrmypdf:v14.4.0
    volumes:
      - ocrmypdf-share:/ocrmypdf:z
    working_dir: /ocrmypdf
    entrypoint: ["bash", "-c"]
    command: ["chmod 777 /ocrmypdf; rm -v /ocrmypdf/*.pdf; while :; do files=(./pre-ocr*); for file in \"$${files[@]}\"; do [ \"$$file\" != \"./pre-ocr*\" ] && ocrmypdf --use-threads --jobs 4 --output-type pdfa --language deu --skip-text --clean --deskew --rotate-pages --rotate-pages-threshold 12.0 \"$$file\" output.pdf || [ $$? -eq 10 ] && rm \"$$file\" && mv output.pdf \"$${file/pre-ocr/post-ocr}\"; done; sleep 1; done"]
```

Basically the PDF is copied to a shared folder. There it is consumed by another container running a more recent version of `ocrmypdf` and replaced with an OCR'd PDF. When the first (and more proper) OCR is done, the file is copied back into the hands of paperless-ngx and the less-proper OCR is done again - resulting somehow in an intact file.

**Comment by ramteid** [comment]
@stumpylog What if we could extend this approach and make it the default way OCR is done in PNGX? There could be a separate `ocrmypdf` microservice providing a (REST) API. This API receives the PDF to OCR and returns an ORC'd PDF. The `webserver` service consumes this API instead of calling `ocrmypdf` natively. 

Actually, this seems to be a somewhat trivial solution for this problem. It would also decouple PNGX from hard-to-update dependencies. What do you think?

**Comment by stumpylog** [comment]
it's an interesting idea, as an opt-in, not the default, so a new parser.  Perhaps based on Alpine, as it is much quicker to provide updated tool versions.  It would also allow hosting on a more resourced machine, for faster OCR.

The problem is two fold:

- How is testing done here
- How is the logic existing in paperless_tesseract ensured to be updated in both?
